src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/ProjectSystemCache.cs

```diff
@@ -547,5 +548,74 @@ public void Dispose()
             Dispose(true);
             GC.SuppressFinalize(this);
         }
+
+        public bool AddProjectRestoreInfoSource(ProjectNames projectNames, object restoreInfoSource)
+        {
+            if (projectNames == null)
+            {
+                throw new ArgumentNullException(nameof(projectNames));
+            }
+
+            if (projectNames.FullName == null)
+            {
+                throw new ArgumentException(
+                    message: string.Format(CultureInfo.CurrentCulture, Strings.PropertyCannotBeNull, nameof(projectNames.FullName)),
+                    paramName: nameof(projectNames));
+            }
+
+            if (restoreInfoSource == null)
+            {
+                throw new ArgumentNullException(nameof(restoreInfoSource));
+            }
+
+            _readerWriterLock.EnterWriteLock();
+
+            try
+            {
+                if (!_projectNamesCache.ContainsKey(projectNames.FullName))
+                {
+                    UpdateProjectNamesCache(projectNames);
+                }
+
+                AddOrUpdateCacheEntry(
+                    projectNames.FullName,
+                    addEntryFactory: k => new CacheEntry
+                    {
+                        ProjectNames = projectNames,
+                        ProjectRestoreInfoSource = restoreInfoSource,
```

Comment:
It looks `ProjectRestoreInfoSource` is `DependencyGraphSpec`, but we're assigning object/IVsProjectRestoreInfoSource type into it. Is it intentional?
https://github.com/NuGet/NuGet.Client/blob/4ba39b98697e715fd09ce762f59c27dc0809eea6/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/ProjectSystemCache.cs#L493

Reply:
ProjectRestoreInfo  vs ProjectRestoreInfoSource. They are 2 different things.